### DISCUSSION TITLE
[Proposal] Add a minimal standalone `python-runtime` formula to speedup Python version migration

### DISCUSSION THREAD

**Discussion by XuehaiPan** [Homebrew/discussions#3274] [Tap maintenance and Homebrew development] [answered]
It has been 7 months past since Python 3.10 was released. The major version of Python 3 in Homebrew/homebrew-core (formula `python@3`) is still Python 3.9.

Listed in https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-988102123 and https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-1126951169. There are lots of formulae depend on `glib` but `glib` has not been migrated yet. This makes our Python 3.10 migration process really slow.

Posted in https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-1127574189, `glib` does not require a specific version of Python. It can be run freely with any Python 3.5+, even with unbrewd ones.

I propose to add a new keg-only formula `python-runtime`. This formula should not allow users to access and do not support installing any `pip` packages. It will be easy to migrate to a new version of Python when 3.11+ is released, because `python-runtime` does not allow installing any `pip` package and for runtime only.

With this formula, we can replace all:

```ruby
depends_on "python@3.9" => [:build, :test]
depends_on "python@3.10" => :build
```

to

```ruby
depends_on "python-runtime" => [:build, :test]
```

And we can have:

```ruby
depends_on "python-runtime"
```

for `glib`.

**Comment by xxyzz** [comment]
Or set Python as build and optional dependency like Arch Linux's [`glib2` package](https://archlinux.org/packages/core/x86_64/glib2/)?

**Reply by XuehaiPan** [reply]
`glib` installs executable `gdbus-codegen` / `glib-genmarshal` / `glib-mkenums` / `gtester-report`. They need a Python interpreter to run.

**Reply by xxyzz** [reply]
That's why Python is make and optional dependency on Arch because only these commands use Python and it can use any version of Python so it won't conflict with other packages' Python version.

**Reply by MikeMcQuaid** [reply]
We don't have optional dependencies.

**Comment by SMillerDev** [comment]
Why not just `uses_from_macos "python"`?

**Reply by XuehaiPan** [reply]
On Linux, `uses_from_macos "python"` will be interpreted as `depends_on "python"`, which is `python@3.9`.

**Reply by SMillerDev** [reply]
So what would `python-runtime` be? Because we can't assume that there is a good python version on Linux.

**Reply by XuehaiPan** [reply]
> So what would python-runtime be?

I think it would be a new formula (`python-runtime.rb`) rather than an alias.

**Reply by SMillerDev** [reply]
Which python version would that be though?

**Reply by XuehaiPan** [reply]
Since Python supports some sort of backward compatibility. We can use as the latest version of Python as we can. It's safe to use the second latest version of Python. It depends on the CI test results. For me, it will be Python 3.9 (second latest) and hopefully 3.10 (latest).

**EDIT:** Note that `python-runtime` is Python 3.9 but it's a completely separated formula with `python@3.9`. It have separated `bin` / `lib` / `include` folders (`$HOMEBREW_PREFIX/opt/python-runtime/xxx` vs. `$HOMEBREW_PREFIX/opt/python@3.9/xxx`).

**Reply by SMillerDev** [reply]
But now you have a `python-runtime` formula that will take months to update because of all the dependencies on it. How does that improve the situation?

**Reply by XuehaiPan** [reply]
1. It can help migrate `glib` immediately.
2. It helps to improve dependency detection. `brew audit` will raise exceptions in the following case:

```
formulaA
|
+-- formulaB
|   |
|   +-- python@3.10
|
+-- glib
|   |
|   +-- python@3.9  # runtime only, it works with python@3.10, but breaks `brew audit` in CI test
...
```

3. It can speed up the migration process when Python 3.11 is released. We just need to update `python-runtime.rb` for `glib`.

**Reply by SMillerDev** [reply]
But that audit is there for a reason. Mixing dependency trees gives issues and it does not matter what you call that dependency.

**Reply by XuehaiPan** [reply]
Suppose now we have:

```
formulaA
|
+-- formulaB
|   |
|   +-- python@3.9
|
+-- glib
|   |
|   +-- python@3.9  # runtime only
...
```

with `python-runtime`, we can easily migrate `formulaB` to `python@3.10`.

```
formulaA
|
+-- formulaB
|   |
|   +-- python@3.10
|
+-- glib
|   |
|   +-- python-runtime
...
```

This is the common situation where the remaining formulae haven't been migrated to Python 3.10 (because of `brew audit` error).

**Reply by SMillerDev** [reply]
But that error would trigger there too. Because that is also a mixed dependency tree.

**Reply by XuehaiPan** [reply]
> But that error would trigger there too. Because that is also a mixed dependency tree.

So I think `python-runtime` should be a standalone formula, rather than a kind of versioned `python` like `python@x.x`. `python@3.9` is conflicted with `python@3.10`, because they are both the versioned formulae of `python`.

**Reply by alexreg** [reply]
What about a special dependency type like `depends_on :python >= 3.5` (syntax may be slightly off), which will look for the following versions of Python, in this order, so long as they satisfy the requested version constraint:
1. OS Python
2. Current `python` formula (whichever version that points to).
3. Latest versioned python formula that is already installed.
4. Latest versioned python formula, full stop

Just an idea.

**Reply by MikeMcQuaid** [reply]
> following versions of Python, in this order, so long as they satisfy the requested version constraint:
> 
> 1. OS Python
> 2. Current `python` formula (whichever version that points to).
> 3. Latest versioned python formula that is already installed.
> 4. Latest versioned python formula, full stop

This is how requirements used to work. It was a testing and reproducibility nightmare, I'm afraid.

**Reply by MikeMcQuaid** [reply]
> So I think `python-runtime` should be a standalone formula, rather than a kind of versioned `python` like `python@x.x`. `python@3.9` is conflicted with `python@3.10`, because they are both the versioned formulae of `python`.

That's just working around the audit here about mixed dependency trees.

The solution to too many Python formulae is not to add another one. If we can figure out how better to split up the migration: let's do that. If not, we need to figure out how to be able to do large migrations like this better (as this also applies to GCC).

**Reply by XuehaiPan** [reply]
> That's just working around the audit here about mixed dependency trees.

Yes, it just helps to migrate those formulae that need a Python interpreter at the building, testing, or runtime. For example `glib` / `vim` / `nodejs` do not install any Python packages by:

```ruby
system "python", *Language::Python.setup_install_args("xxx")
```

These formulae only use the Python standard libraries.

> The solution to too many Python formulae is not to add another one.

For many Python formulae (PyPI packages, e.g. `numpy` / `pylint` / `ipython`), they will install packages to `lib/site-packages`, this is out of the scope the proposed `python-runtime` can handle. Besides, PyPI packages have their own resource dependencies (the `resource "..." do ... end` block in `.rb` file) and the Python version dependency. We should explicitly have `depends_on "python@3.x"` rather than an unversioned `python-runtime`.

**Reply by XuehaiPan** [reply]
> If we can figure out how better to split up the migration: let's do that.

Rerun the code I posted in https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-988102123

Now we have to migrate 121 formulae (family 3) all together in a single PR.

```
HEAD: 4bc9ae2ce6d001aff244820c7d71c474142838fd
```

- **1. stand-alones (39 formulae / 39 PR)**

    ```
    [['astrometry-net', 'awscli@1', 'binwalk', 'blastem', 'csound', 'dnsviz', 'eralchemy', 'eye-d3', 'fobis', 'folderify', 'ford', 'genometools', 'gitfs', 'gprof2dot', 'hy', 'infer', 'jdtls', 'jupyterlab', 'ldns', 'libstfl', 'libtensorflow@1', 'litani', 'llnode', 'lv2', 'mapproxy', 'mavsdk', 'mkvtomp4', 'node@12', 'open-babel', 'pyenv', 'q', 'simgrid', 'soapysdr', 'statik', 'subliminal', 'tbb@2020', 'woob', 'wxpython', 'yle-dl']]
    ```

- **2. Family `cython` (3 formulae / 1 PR)**

    ```
    [['cython'], ['vapoursynth'], ['mpv']]
    ```

- **3. Family `rust` (121 formulae / 1 PR)**

    ```
    [['rust', 'python-tabulate', 'glib', 'libxml2', 'py3cairo', 'sip', 'boost-python3', 'tbb', 'libcython', 'gpgme', 'llvm@12', 'apache-arrow', 'volk', 'opencolorio', 's3cmd', 'pythran', 'uhd', 'cryptominisat', 'wllvm', 'texlive'], ['gobject-introspection', 'itstool', 'pyqt-builder', 'scipy', 'gdal', 'fontforge', 'stp', 'notmuch', 'libsigrokdecode', 'spidermonkey@78', 'c7n', 'homeassistant-cli', 'aws-google-auth', 'litecli', 'pgcli', 'root', 'pyflow', 'snapcraft', 'regipy', 's3ql', 'gtk-doc', 'athenacli', 'eigenpy', 'spidermonkey', 'opencv@3', 'fdroidserver', 'suricata', 'ansible@2.9', 'sqlite-utils', 'esphome', 'duplicity', 'todoman', 'profanity', 'trezor-agent', 'localstack', 'checkov', 'lcm', 'openimageio', 'charmcraft', 'libtorrent-rasterbar', 'dvc', 'vsts-cli', 'ocrmypdf', 'lensfun', 'ghc', 'anime-downloader', 'rdkit', 'mycli', 'platformio', 'awscli'], ['pygobject3', 'at-spi2-core', 'spice-gtk', 'klee', 'anjuta', 'evince', 'mftrace', 'networkit', 'easy-tag', 'alot', 'gucharmap', 'mapnik', 'lilypond', 'gst-devtools', 'yelp-tools', 'mapserver', 'gupnp', 'gom'], ['qt@5', 'qt', 'zim', 'graph-tool', 'gst-python', 'nicotine-plus', 'pympress', 'pytouhou', 'libpeas', 'terminator', 'gexiv2', 'virt-manager', 'xdot', 'mat2', 'at-spi2-atk', 'libsigrok'], ['pyqt@5', 'pyqt', 'pyside', 'pushpin'], ['vtk', 'vtk@8.2', 'pyqt-3d', 'gnuradio', 'pyqt-networkauth', 'carla', 'qscintilla2', 'git-cola', 'urh', 'trimage'], ['opencv', 'gdcm'], ['openalpr']]
    ```

**Reply by XuehaiPan** [reply]
After adding `python-runtime` formula and changing the runtime-only Python dependencies. We can split the huge dependency tree into multiple smaller ones.

```
HEAD: 4bc9ae2ce6d001aff244820c7d71c474142838fd
```

- **1. stand-alones (61 formulae / 61 PR)**

    ```
    [['ansible@2.9', 'astrometry-net', 'at-spi2-atk', 'awscli', 'awscli@1', 'binwalk', 'blastem', 'csound', 'dnsviz', 'duplicity', 'eralchemy', 'eye-d3', 'fobis', 'folderify', 'ford', 'genometools', 'gitfs', 'gom', 'gprof2dot', 'gst-devtools', 'gtk-doc', 'gupnp', 'hy', 'infer', 'jdtls', 'jupyterlab', 'lcm', 'ldns', 'lensfun', 'libsigrokdecode', 'libstfl', 'libtensorflow@1', 'litani', 'llnode', 'lv2', 'mapproxy', 'mavsdk', 'mkvtomp4', 'node@12', 'ocrmypdf', 'open-babel', 'pushpin', 'pyenv', 'pyflow', 'pyside', 'q', 's3ql', 'simgrid', 'snapcraft', 'soapysdr', 'spidermonkey', 'spidermonkey@78', 'statik', 'subliminal', 'suricata', 'tbb@2020', 'trezor-agent', 'vsts-cli', 'woob', 'wxpython', 'yle-dl']]
    ```

- **2. Family `s3cmd` (2 formulae / 1 PR)**

    ```
    [['s3cmd'], ['fdroidserver']]
    ```

- **3. Family `cython` (3 formulae / 1 PR)**

    ```
    [['cython'], ['vapoursynth'], ['mpv']]
    ```

- **4. Family `gdal` (3 formulae / 1 PR)**

    ```
    [['gdal'], ['mapserver', 'mapnik']]
    ```

- **5. Family `fontforge` (4 formulae / 1 PR)**

    ```
    [['fontforge', 'texlive'], ['lilypond', 'mftrace']]
    ```

- **6. Family `gpgme` (4 formulae / 1 PR)**

    ```
    [['gpgme', 'notmuch'], ['profanity', 'alot']]
    ```

- **7. Family `itstool` (6 formulae / 1 PR)**

    ```
    [['itstool'], ['yelp-tools', 'easy-tag', 'gucharmap', 'evince', 'anjuta']]
    ```

- **8. Family `python-tabulate` (24 formulae / 1 PR)**

    ```
    [['python-tabulate', 'apache-arrow', 'cryptominisat', 'llvm@12', 'wllvm'], ['stp', 'sqlite-utils', 'homeassistant-cli', 'regipy', 'mycli', 'aws-google-auth', 'anime-downloader', 'litecli', 'c7n', 'todoman', 'esphome', 'charmcraft', 'checkov', 'localstack', 'pgcli', 'athenacli', 'dvc', 'platformio'], ['klee']]
    ```

- **9. Family `py3cairo` (47 formulae / 1 PR)**

    ```
    [['py3cairo', 'sip', 'boost-python3', 'tbb', 'libcython', 'opencolorio', 'uhd', 'pythran', 'volk'], ['pygobject3', 'pyqt-builder', 'scipy', 'rdkit', 'eigenpy', 'libtorrent-rasterbar', 'root', 'openimageio', 'opencv@3'], ['pyqt@5', 'pyqt', 'xdot', 'networkit', 'zim', 'nicotine-plus', 'virt-manager', 'gexiv2', 'pytouhou', 'graph-tool', 'libpeas', 'terminator', 'gst-python', 'pympress', 'mat2', 'libsigrok'], ['vtk', 'vtk@8.2', 'urh', 'qscintilla2', 'git-cola', 'gnuradio', 'pyqt-networkauth', 'trimage', 'carla', 'pyqt-3d'], ['opencv', 'gdcm'], ['openalpr']]
    ```

**Reply by MikeMcQuaid** [reply]
> Now we have to migrate 121 formulae (family 3) all together in a single PR.

Yes, I think this is what we should do.

**Comment by XuehaiPan** [comment]
Potential issues for `python-runtime`:

1. I'm not sure whether it will break `which("python3")` or not in downstream formulae's `install` block.

2. The shebang in downstream formula bottles should be

   ```bash
   #!@@HOMEBREW_PREFIX@@/opt/python-runtime/bin/python3
   ```

   rather than

   ```bash
   #!@@HOMEBREW_PREFIX@@/opt/python-runtime/bin/python3.x
   ```

   Because `python-runtime` is unversioned.

**Reply by SMillerDev** [reply]
And what would it point to? Because centOS 7 (supported OS for Linuxbrew) does not have python 3 by default AFAIK

**Reply by XuehaiPan** [reply]
If we have `depends_on "python@3.10"` in the formula script. Then `which("python3")` will give `"${HOMEBREW_PREFIX}/opt/python@3.10/bin/python3"` at building time.

**Comment by iMichka** [comment]
I would be rather in favour of investing time splitting the glib formula in two, like done in conda: https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-1127014411 / https://uwekorn.com/2020/10/19/r-without-python.html
One with and one without Python.

This should help speed up the migration

**Reply by XuehaiPan** [reply]
> This should help speed up the migration

Another benefit for a runtime-only Python is we can have:

```ruby
depends_on "python-runtime" => [:build, :test]
```

rather than

```ruby
depends_on "python@3.x" => [:build, :test]
```

The latter one needs additional migration efforts when a new major version of Python is released.

**Reply by SMillerDev** [reply]
So how would you update `python-runtime` to a newer python version?

**Reply by XuehaiPan** [reply]
Since `python-runtime` is the second latest version of Python, just copy the content from `python@3.10.rb` and delete the corresponding part for `pip` when `python@3.11` is released.

**Comment by alexreg** [comment]
Ah, fair enough then.

**Comment by iMichka** [comment, marked as answer]
> Now we have to migrate 121 formulae (family 3) all together in a single PR.
> Yes, I think this is what we should do.

Agreed. 121 is small enough for a single PR.

Again, we could reduce that number by working on:
https://github.com/Homebrew/homebrew-core/pull/99064 (removing python-tabulate out of the picture)
https://github.com/Homebrew/homebrew-core/pull/87075#issuecomment-1127014411 / https://uwekorn.com/2020/10/19/r-without-python.html (removing glib out of the picture)
